File: src/cappuccino/ide/intellij/plugin/formatting/ObjJIndentProcessor.java
package cappuccino.ide.intellij.plugin.formatting;

import cappuccino.ide.intellij.plugin.parser.ObjJParserDefinition;
import cappuccino.ide.intellij.plugin.psi.utils.ObjJTreeUtilKt;
import com.intellij.formatting.FormattingMode;
import com.intellij.formatting.Indent;
import com.intellij.lang.ASTNode;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.TokenType;
import com.intellij.psi.codeStyle.CommonCodeStyleSettings;
import com.intellij.psi.formatter.FormatterUtil;
import com.intellij.psi.tree.IElementType;
import com.intellij.psi.tree.TokenSet;
import org.jetbrains.annotations.NotNull;

import static cappuccino.ide.intellij.plugin.psi.types.ObjJTypes.*;
import static com.intellij.psi.TokenType.WHITE_SPACE;

public class ObjJIndentProcessor {

    public static final TokenSet EXPRESSIONS = TokenSet
            .create(ObjJ_ARRAY_INDEX_SELECTOR, ObjJ_VARIABLE_DECLARATION, ObjJ_FUNCTION_CALL, ObjJ_LEFT_EXPR, ObjJ_RIGHT_EXPR, ObjJ_EXPR);

    public static final TokenSet BLOCKS = TokenSet.create(ObjJ_BLOCK_ELEMENT, ObjJ_BRACKET_LESS_BLOCK, ObjJ_METHOD_BLOCK, ObjJ_PROTOCOL_SCOPED_BLOCK,
            ObjJ_STATEMENT_OR_BLOCK);

    public static final TokenSet COMMENTS = ObjJParserDefinition.Companion.getCOMMENTS();

    public static final TokenSet CLASS_DECLARATIONS = TokenSet.create(ObjJ_IMPLEMENTATION_DECLARATION, ObjJ_PROTOCOL_DECLARATION);

    private final CommonCodeStyleSettings settings;

    public ObjJIndentProcessor(CommonCodeStyleSettings settings) {
        this.settings = settings;
    }

    public Indent getChildIndent(final ASTNode node, final FormattingMode mode) {
        final IElementType elementType = node.getElementType();
        final ASTNode prevSibling = ObjJTreeUtilKt.getPreviousNonEmptySiblingIgnoringComments(node);
        final IElementType prevSiblingType = prevSibling == null ? null : prevSibling.getElementType();
        final ASTNode parent = node.getTreeParent();
        final IElementType parentType = parent != null ? parent.getElementType() : null;
        final ASTNode superParent = parent == null ? null : parent.getTreeParent();
        final IElementType superParentType = superParent == null ? null : superParent.getElementType();

        final int braceStyle = settings.BRACE_STYLE;

        if (parent == null || parent.getTreeParent() == null/* || parentType == EMBEDDED_CONTENT*/) {
            return Indent.getNoneIndent();
        }
        if (elementType == ObjJ_BLOCK_COMMENT_BODY) {
            return Indent.getContinuationIndent();
        }
        if (elementType == ObjJ_BLOCK_COMMENT_LEADING_ASTERISK || elementType == ObjJ_BLOCK_COMMENT_END) {
            return Indent.getSpaceIndent(1, true);
        }
        if (settings.KEEP_FIRST_COLUMN_COMMENT && (elementType == ObjJ_SINGLE_LINE_COMMENT || elementType == ObjJ_BLOCK_COMMENT)) {
            final ASTNode previousNode = node.getTreePrev();
            if (previousNode != null && previousNode.getElementType() == WHITE_SPACE && previousNode.getText().endsWith("\n")) {
                return Indent.getAbsoluteNoneIndent();
            }
        }

        if (COMMENTS.contains(elementType) && prevSiblingType == ObjJ_OPEN_BRACE && CLASS_DECLARATIONS.contains(parentType)) {
            return Indent.getNormalIndent();
        }

        /*if (parentType == ENUM_DEFINITION && isBetweenBraces(node)) {
            // instead of isBetweenBraces(node) we can parse enum block as a separate ASTNode, or build formatter blocks not tied to AST.
            return Indent.getNormalIndent();
        }*/

        if (parentType == ObjJ_ARRAY_LITERAL || parentType == ObjJ_OBJECT_LITERAL) {
            if (elementType == ObjJ_OPEN_BRACE ||
                    elementType == ObjJ_AT_OPEN_BRACE ||
                    elementType == ObjJ_CLOSE_BRACE ||
                    elementType == ObjJ_OPEN_BRACKET ||
                    elementType == ObjJ_AT_OPENBRACKET ||
                    elementType == ObjJ_CLOSE_BRACKET
            ) {
                return Indent.getNoneIndent();
            }
            // Be careful to preserve typing behavior.
            if (elementType == ObjJ_PROPERTY_ASSIGNMENT || elementType == ObjJ_EXPR || elementType == ObjJ_COMMA) {
                return Indent.getNormalIndent();
            }
            if (COMMENTS.contains(elementType)) {
                return Indent.getNormalIndent();
            }
            return Indent.getNoneIndent();
        }

        if (elementType == ObjJ_OPEN_BRACE || elementType == ObjJ_CLOSE_BRACE) {
            switch (braceStyle) {
                case CommonCodeStyleSettings.END_OF_LINE:
                    if (elementType == ObjJ_OPEN_BRACE && FormatterUtil.isPrecededBy(parent, ObjJ_SINGLE_LINE_COMMENT, WHITE_SPACE)) {
                        // Use Nystrom style rather than Allman.
                        return Indent.getContinuationIndent();
                    } // FALL THROUGH
                case CommonCodeStyleSettings.NEXT_LINE:
                case CommonCodeStyleSettings.NEXT_LINE_IF_WRAPPED:
                    return Indent.getNoneIndent();
                case CommonCodeStyleSettings.NEXT_LINE_SHIFTED:
                case CommonCodeStyleSettings.NEXT_LINE_SHIFTED2:
                    return Indent.getNormalIndent();
                default:
                    return Indent.getNoneIndent();
            }
        }

        if (parentType == PARENTHESIZED_EXPRESSION) {
            if (elementType == LPAREN || elementType == RPAREN) {
                return Indent.getNoneIndent();
            }
            return Indent.getContinuationIndent();
        }

        if (elementType == CLASS_MEMBERS) {
            return Indent.getNormalIndent();
        }
        if (BLOCKS.contains(parentType)) {
            final PsiElement psi = node.getPsi();
            if (psi.getParent() instanceof PsiFile) {
                return Indent.getNoneIndent();
            }
            return Indent.getNormalIndent();
        }
        if (elementType == LPAREN && (superParentType == METADATA || parentType == ARGUMENTS)) {
            return Indent.getNormalIndent();
        }
        if (parentType == ARGUMENTS) {
            if (COMMENTS.contains(elementType)) {
                return Indent.getNormalIndent();
            }
            return Indent.getNoneIndent();
        }
        if (parentType == ARGUMENT_LIST) {
            // see https://github.com/ObjJ-lang/ObjJ_style/issues/551
            return parent.getLastChildNode().getElementType() == COMMA ? Indent.getNormalIndent() : Indent.getContinuationIndent();
        }
        if (parentType == FORMAL_PARAMETER_LIST || parentType == PARAMETER_TYPE_LIST) {
            return Indent.getContinuationIndent();
        }
        if (parentType == OPTIONAL_FORMAL_PARAMETERS &&
                elementType != LBRACE && elementType != RBRACE &&
                elementType != LBRACKET && elementType != RBRACKET) {
            return Indent.getNormalIndent();
        }
        if (parentType == FOR_STATEMENT && prevSiblingType == FOR_LOOP_PARTS_IN_BRACES && !BLOCKS.contains(elementType)) {
            return Indent.getNormalIndent();
        }
        if (parentType == SWITCH_STATEMENT && (elementType == SWITCH_CASE || elementType == DEFAULT_CASE)) {
            return Indent.getNormalIndent();
        }
        if ((parentType == SWITCH_CASE || parentType == DEFAULT_CASE) && elementType == STATEMENTS) {
            return Indent.getNormalIndent();
        }
        if (parentType == WHILE_STATEMENT && prevSiblingType == RPAREN && !BLOCKS.contains(elementType)) {
            return Indent.getNormalIndent();
        }
        if (parentType == DO_WHILE_STATEMENT && prevSiblingType == DO && !BLOCKS.contains(elementType)) {
            return Indent.getNormalIndent();
        }
        if ((parentType == RETURN_STATEMENT) &&
                prevSiblingType == RETURN &&
                !BLOCKS.contains(elementType)) {
            return Indent.getNormalIndent();
        }
        if (parentType == IF_STATEMENT && !BLOCKS.contains(elementType) &&
                (prevSiblingType == RPAREN || (prevSiblingType == ELSE && elementType != IF_STATEMENT))) {
            return Indent.getNormalIndent();
        }
        if (elementType == CASCADE_REFERENCE_EXPRESSION) {
            return Indent.getNormalIndent();
        }
        if (elementType == OPEN_QUOTE && prevSiblingType == CLOSING_QUOTE && parentType == STRING_LITERAL_EXPRESSION) {
            return Indent.getContinuationIndent();
        }
        if (BINARY_EXPRESSIONS.contains(parentType) && prevSibling != null) {
            return Indent.getContinuationIndent();
        }
        if (elementType == COLON || parentType == TERNARY_EXPRESSION && elementType == QUEST) {
            return Indent.getContinuationIndent();
        }
        if (elementType == HIDE_COMBINATOR || elementType == SHOW_COMBINATOR) {
            return Indent.getContinuationIndent();
        }
        if (parentType == FUNCTION_BODY) {
            if (FormatterUtil.isPrecededBy(node, EXPRESSION_BODY_DEF)) {
                return Indent.getContinuationIndent();
            }
        }
        if (elementType == CALL_EXPRESSION) {
            if (FormatterUtil.isPrecededBy(node, EXPRESSION_BODY_DEF)) {
                return Indent.getContinuationIndent();
            }
            if (FormatterUtil.isPrecededBy(node, ASSIGNMENT_OPERATOR)) {
                return Indent.getContinuationIndent();
            }
        }
        if ((elementType == REFERENCE_EXPRESSION || BINARY_EXPRESSIONS.contains(elementType)) &&
                (FormatterUtil.isPrecededBy(node, ASSIGNMENT_OPERATOR) || FormatterUtil.isPrecededBy(node, EQ))) {
            return Indent.getContinuationIndent();
        }
        if (elementType == VAR_DECLARATION_LIST_PART) {
            return Indent.getContinuationIndent();
        }

        if (elementType == SUPER_CALL_OR_FIELD_INITIALIZER) {
            return Indent.getContinuationIndent();
        }
        if (parentType == SUPER_CALL_OR_FIELD_INITIALIZER && elementType != COLON) {
            return Indent.getNormalIndent();
        }

        if (parentType == CLASS_DEFINITION) {
            if (elementType == SUPERCLASS || elementType == INTERFACES || elementType == MIXINS) {
                return Indent.getContinuationIndent();
            }
        }
        if (parentType == MIXIN_APPLICATION && elementType == MIXINS) {
            return Indent.getContinuationIndent();
        }

        if (parentType == LIBRARY_NAME_ELEMENT) {
            return Indent.getContinuationIndent();
        }

        if (elementType == SEMICOLON && FormatterUtil.isPrecededBy(node, SINGLE_LINE_COMMENT, WHITE_SPACE)) {
            return Indent.getContinuationIndent();
        }

        if (elementType == DOT || elementType == QUEST_DOT) {
            return Indent.getContinuationIndent();
        }

        if (parentType == TYPE_LIST && elementType == TYPE) {
            return Indent.getContinuationIndent();
        }

        if (elementType == OPEN_QUOTE && parentType == STRING_LITERAL_EXPRESSION && superParentType == VAR_INIT) {
            if (node.getText().length() < 3) {
                return Indent.getContinuationIndent();
            }
        }

        if (elementType == RAW_SINGLE_QUOTED_STRING && parentType == STRING_LITERAL_EXPRESSION && superParentType == VAR_INIT) {
            return Indent.getContinuationIndent();
        }

        if (parentType == LONG_TEMPLATE_ENTRY && EXPRESSIONS.contains(elementType)) {
            return Indent.getContinuationIndent();
        }
        return Indent.getNoneIndent();
    }

    private static boolean isBetweenBraces(@NotNull final ASTNode node) {
        final IElementType elementType = node.getElementType();
        if (elementType == LBRACE || elementType == RBRACE) return false;

        for (ASTNode sibling = node.getTreePrev(); sibling != null; sibling = sibling.getTreePrev()) {
            if (sibling.getElementType() == LBRACE) return true;
        }

        return false;
    }
}

File: src/cappuccino/ide/intellij/plugin/formatting/ObjJIndentProcessor.kt
package cappuccino.ide.intellij.plugin.formatting

import cappuccino.ide.intellij.plugin.psi.*
import cappuccino.ide.intellij.plugin.psi.types.ObjJTokenSets
import cappuccino.ide.intellij.plugin.psi.utils.*
import com.intellij.formatting.Indent
import com.intellij.lang.ASTNode
import com.intellij.psi.PsiFile
import com.intellij.psi.codeStyle.CommonCodeStyleSettings
import com.intellij.psi.formatter.FormatterUtil
import com.intellij.psi.tree.TokenSet

import cappuccino.ide.intellij.plugin.psi.types.ObjJTypes.*
import cappuccino.ide.intellij.plugin.settings.ObjJCodeStyleSettings
import cappuccino.ide.intellij.plugin.stubs.types.ObjJStubTypes
import com.intellij.psi.TokenType.WHITE_SPACE
import java.util.logging.Logger

class ObjJIndentProcessor(private val settings: CommonCodeStyleSettings, private val objjSettings:ObjJCodeStyleSettings) {
    fun getChildIndent(node: ASTNode): Indent? {
        val elementType = node.elementType
        val prevSibling = node.getPreviousNonEmptySiblingIgnoringComments()
        val prevSiblingType = prevSibling?.elementType
        val nextSibling = node.getNextNonEmptyNode(true)
        val nextSiblingType = nextSibling?.elementType
        val parent = node.treeParent
        val parentType = parent?.elementType
        val superParent = parent?.treeParent
        val superParentType = superParent?.elementType

        if (parentType == ObjJStubTypes.FILE) {
            return Indent.getNoneIndent()
        }

        if (elementType == ObjJ_OPEN_BRACE || elementType == ObjJ_CLOSE_BRACE) {
            if (node.treeParent == ObjJ_VARIABLE_DECLARATION ||
                    node.treeParent?.treeParent == ObjJ_VARIABLE_DECLARATION ||
                    node.treeParent?.treeParent?.treeParent == ObjJ_VARIABLE_DECLARATION) {
                return Indent.getNormalIndent()
            }
            return Indent.getNoneIndent()
        }

        if (prevSibling == ObjJ_IF_STATEMENT || prevSibling == ObjJ_ELSE_IF_STATEMENT || prevSibling == ObjJ_STATEMENT_OR_BLOCK)
            Indent.getNoneIndent()

        if (prevSibling == ObjJ_OPEN_BRACE || nextSiblingType == ObjJ_CLOSE_BRACE) {
            Indent.getNormalIndent()
        }

        if (parentType == ObjJ_INSTANCE_VARIABLE_LIST) {
            if (elementType == ObjJ_OPEN_BRACE || elementType == ObjJ_CLOSE_BRACE)
                return Indent.getNoneIndent()
            return Indent.getNormalIndent()
        }


        if (parentType == ObjJ_FOR_STATEMENT || parentType == ObjJ_WHILE_STATEMENT || parentType == ObjJ_DO_WHILE_STATEMENT) {
            if (elementType == ObjJ_STATEMENT_OR_BLOCK)
                return Indent.getNoneIndent()
            if (elementType == ObjJ_STATEMENT_OR_BLOCK && node.firstChildNode?.elementType !in ObjJTokenSets.INDENT_CHILDREN) {
                return Indent.getNormalIndent()
            }
            return Indent.getNoneIndent()
        } else if (parentType == ObjJ_STATEMENT_OR_BLOCK) {
            return Indent.getNoneIndent()
        }

        if (elementType in ObjJTokenSets.INDENT_CHILDREN)
            return Indent.getNoneIndent()

        if (elementType == ObjJ_STATEMENT_OR_BLOCK) {
            return Indent.getNoneIndent()
        }

        if (elementType == ObjJ_RIGHT_EXPR)
            return Indent.getContinuationWithoutFirstIndent()

        if (parentType in ObjJTokenSets.INDENT_CHILDREN) {
            if (elementType == ObjJ_OPEN_BRACE || elementType == ObjJ_CLOSE_BRACE)
                return Indent.getNoneIndent()
            return Indent.getNormalIndent()
        }

        if (parentType == ObjJ_IMPLEMENTATION_DECLARATION) {
            if ((prevSibling == ObjJ_OPEN_BRACE || prevSibling == ObjJ_SEMI_COLON || prevSibling == ObjJ_INSTANCE_VARIABLE_DECLARATION) && elementType != ObjJ_CLOSE_BRACE) {
                return Indent.getNormalIndent()
            }
        }

        if (parent == null || parent.treeParent == null) {
            return Indent.getNoneIndent()
        }

        if ((parentType == ObjJ_PROTOCOL_DECLARATION || parentType == ObjJ_PROTOCOL_SCOPED_METHOD_BLOCK) && elementType == ObjJ_METHOD_HEADER) {
            return Indent.getNoneIndent()
        }

        if (elementType == ObjJ_BLOCK_COMMENT_BODY) {
            return Indent.getContinuationIndent()
        }

        if (elementType == ObjJ_BLOCK_COMMENT_LEADING_ASTERISK || elementType == ObjJ_BLOCK_COMMENT_END) {
            return Indent.getSpaceIndent(1, true)
        }

        if (settings.KEEP_FIRST_COLUMN_COMMENT && (elementType == ObjJ_SINGLE_LINE_COMMENT || elementType == ObjJ_BLOCK_COMMENT)) {
            val previousNode = node.treePrev
            if (previousNode != null && previousNode.elementType == WHITE_SPACE && previousNode.text.endsWith("\n")) {
                return Indent.getAbsoluteNoneIndent()
            }
        }

        if (parentType == ObjJ_LOGIC_EXPR_PRIME) {
            return Indent.getContinuationIndent()
        }

        if (ObjJTokenSets.COMMENTS.contains(elementType) && prevSiblingType == ObjJ_OPEN_BRACE && ObjJTokenSets.CLASS_DECLARATIONS.contains(parentType)) {
            return Indent.getNormalIndent()
        }

        if (parentType == ObjJ_ARRAY_LITERAL || parentType == ObjJ_OBJECT_LITERAL) {
            if (elementType == ObjJ_OPEN_BRACE ||
                    elementType == ObjJ_AT_OPEN_BRACE ||
                    elementType == ObjJ_CLOSE_BRACE) {
                return Indent.getNoneIndent()
            }

            if (elementType == ObjJ_PROPERTY_ASSIGNMENT) {
                if (objjSettings.ALIGN_PROPERTIES) {
                    val propertyAssignment =
                            node.psi?.getSelfOrParentOfType(ObjJPropertyAssignment::class.java)
                                    ?: return Indent.getNormalIndent()
                    val spacing = propertyAssignment.getObjectLiteralPropertySpacing()
                    if (spacing != null && spacing > 0) {
                        return Indent.getSpaceIndent(spacing)
                    }
                }
                return Indent.getNormalIndent()
            }

            // Be careful to preserve typing behavior.
            if (elementType == ObjJ_PROPERTY_ASSIGNMENT || elementType == ObjJ_COMMA) {
                return Indent.getNormalIndent()
            }
            if (elementType == ObjJ_EXPR)
                return Indent.getContinuationIndent()
            return if (ObjJTokenSets.COMMENTS.contains(elementType)) {
                Indent.getNormalIndent()
            } else Indent.getNoneIndent()
        }

        if (parentType == ObjJ_METHOD_BLOCK)
            return Indent.getNormalIndent()
        if (parentType == ObjJ_VARIABLE_DECLARATION_LIST) {
            return Indent.getSpaceIndent(4)
        }


        if (elementType == ObjJ_PREPROCESSOR_IF_STATEMENT) {
            if (parentType == ObjJ_METHOD_BLOCK) {
                //return Indent.getNoneIndent()
            }
            return Indent.getNormalIndent(false)
        }

        if (parentType in ObjJTokenSets.INDENT_CHILDREN) {
            val psi = node.psi
            return if (psi.parent is PsiFile) {
                Indent.getNoneIndent()
            } else Indent.getNormalIndent()
        }

        if (elementType == ObjJ_METHOD_DECLARATION) {
            return Indent.getNoneIndent()
        }


        if (elementType in ObjJTokenSets.METHOD_HEADER_DECLARATION_SELECTOR) {
            if (objjSettings.ALIGN_SELECTORS_IN_METHOD_DECLARATION) {
                val selectorSpacing = node.psi?.getSelfOrParentOfType(ObjJMethodDeclarationSelector::class.java)?.getSelectorAlignmentSpacing(objjSettings)
                //LOGGER.info("Get Method Dec Selector Indent Spacing: $selectorSpacing")
                if (selectorSpacing != null && selectorSpacing > 0) {
                    return Indent.getSpaceIndent(selectorSpacing)
                } else {
                    //LOGGER.info("Failed to get selector spacing for: ${node.text}")
                }
            }
            return Indent.getContinuationIndent()
        }

        if (parentType == ObjJ_PAREN_ENCLOSED_EXPR || parentType == ObjJ_ARRAY_INDEX_SELECTOR) {
            return if (elementType == ObjJ_OPEN_PAREN || elementType == ObjJ_CLOSE_PAREN || elementType == ObjJ_OPEN_BRACKET || elementType == ObjJ_CLOSE_BRACKET) {
                Indent.getNoneIndent()
            } else Indent.getContinuationIndent()
        }

        if (parentType == ObjJ_FOR_STATEMENT && (prevSiblingType == ObjJ_FOR_LOOP_PARTS_IN_BRACES || prevSiblingType == ObjJ_FOR_LOOP_HEADER)) {
            if (elementType !in ObjJTokenSets.INDENT_CHILDREN)
                return Indent.getNormalIndent()
            return Indent.getNoneIndent()
        }

        if (elementType == ObjJ_FOR_LOOP_PARTS_IN_BRACES || elementType == ObjJ_FOR_LOOP_HEADER) {
            return Indent.getNoneIndent()
        }

        if (parentType == ObjJ_IN_EXPR) {
            return when {
                elementType == ObjJ_IN -> Indent.getContinuationIndent()
                elementType == ObjJ_OPEN_PAREN -> Indent.getNoneIndent()
                prevSiblingType == ObjJ_OPEN_PAREN -> Indent.getNormalIndent()
                elementType == ObjJ_EXPR -> Indent.getContinuationIndent()
                elementType == ObjJ_OPEN_PAREN || elementType == ObjJ_CLOSE_PAREN -> Indent.getNoneIndent()
                else -> Indent.getNoneIndent()
            }
        }

        if (parentType == ObjJ_SWITCH_STATEMENT && (elementType == ObjJ_CASE_CLAUSE || elementType == ObjJ_DEFAULT_CLAUSE)) {
            return Indent.getNormalIndent()
        }

        if ((parentType == ObjJ_CASE_CLAUSE || parentType == ObjJ_DEFAULT_CLAUSE) && elementType == ObjJ_BRACKET_LESS_BLOCK) {
            return Indent.getNoneIndent()
        }

        if (parentType == ObjJ_WHILE_STATEMENT && prevSiblingType == ObjJ_CLOSE_PAREN && !ObjJTokenSets.INDENT_CHILDREN.contains(elementType)) {
            return Indent.getNormalIndent()
        }

        if (parentType == ObjJ_DO_WHILE_STATEMENT && prevSiblingType == ObjJ_DO && !ObjJTokenSets.INDENT_CHILDREN.contains(elementType)) {
            return Indent.getNoneIndent()
        }

        if (parentType == ObjJ_RETURN_STATEMENT &&
                prevSiblingType == ObjJ_RETURN &&
                !ObjJTokenSets.INDENT_CHILDREN.contains(elementType)) {
            return Indent.getNoneIndent()
        }

        if (parentType == ObjJ_IF_STATEMENT && !ObjJTokenSets.INDENT_CHILDREN.contains(elementType) &&
                (prevSiblingType == ObjJ_CLOSE_PAREN || (prevSiblingType == ObjJ_ELSE && elementType != ObjJ_IF_STATEMENT))) {
            return Indent.getNormalIndent()
        }

        if (parentType == ObjJ_PREPROCESSOR_BODY_STATEMENTS || elementType == ObjJ_PREPROCESSOR_BODY_STATEMENTS) {
            return Indent.getNormalIndent()
        }

        if (elementType  == ObjJ_PREPROCESSOR_IF_CONDITION) {
            return Indent.getNoneIndent()
        }

        /*
        if (elementType == ObjJ_OPEN_QUOTE && prevSiblingType == CLOSING_QUOTE && parentType == STRING_LITERAL_EXPRESSION) {
            return Indent.getContinuationIndent();
        }*/


        if (elementType == ObjJ_FUNCTION_CALL) {
            if (FormatterUtil.isPrecededBy(node, ObjJ_ASSIGNMENT_OPERATOR, WHITE_SPACE)) {
                return Indent.getContinuationIndent()
            }
        }

        if (parentType == ObjJ_INHERITED_PROTOCOL_LIST) {
            return if (elementType == ObjJ_LESS_THAN || elementType == ObjJ_GREATER_THAN)
                Indent.getNoneIndent()
            else
                Indent.getNormalIndent()
        }



        if (elementType == ObjJ_QUALIFIED_METHOD_CALL_SELECTOR) {
            if (objjSettings.ALIGN_SELECTORS_IN_METHOD_CALL) {
                val selectorSpacing = node.psi?.getSelfOrParentOfType(ObjJQualifiedMethodCallSelector::class.java)?.getSelectorAlignmentSpacing(objjSettings.ALIGN_FIRST_SELECTOR_IN_METHOD_CALL)
                if (selectorSpacing != null && selectorSpacing >= 4) {
                    return Indent.getSpaceIndent(selectorSpacing)
                }
            }
            return Indent.getNormalIndent()
        }

        if (parentType == ObjJ_QUALIFIED_METHOD_CALL_SELECTOR) {
            if (FormatterUtil.isPrecededBy(node, ObjJ_COLON, WHITE_SPACE)) {
                return Indent.getContinuationIndent()
            }
        }

        if (parentType == ObjJ_METHOD_CALL) {
            if (elementType == ObjJ_CALL_TARGET)
                return Indent.getNormalIndent()
            if (FormatterUtil.isPrecededBy(node, ObjJ_CALL_TARGET, WHITE_SPACE)) {
                return Indent.getNormalIndent()
            }
            if (elementType == ObjJ_COLON)
                return Indent.getContinuationIndent()
            if (elementType == ObjJ_OPEN_BRACKET || elementType == ObjJ_CLOSE_BRACKET) {
                return Indent.getNoneIndent()
            }
        }

        if (elementType == ObjJ_METHOD_HEADER_SELECTOR_FORMAL_VARIABLE_TYPE || parentType == ObjJ_METHOD_HEADER_SELECTOR_FORMAL_VARIABLE_TYPE) {
            return Indent.getNormalIndent()
        }

        if (parentType == ObjJ_ASSIGNMENT_EXPR_PRIME) {
            return Indent.getContinuationIndent()
        }

        if (ObjJTokenSets.CLASS_DECLARATIONS.contains(parentType)) {
            return Indent.getNoneIndent()
        }

        if (parentType == ObjJ_IMPORT_FILE || parentType == ObjJ_IMPORT_FRAMEWORK) {
            if (FormatterUtil.isPrecededBy(node, ObjJ_AT_IMPORT, WHITE_SPACE)) {
                return Indent.getContinuationIndent()
            }
        }

        if (elementType == ObjJ_SEMI_COLON && FormatterUtil.isPrecededBy(node, ObjJ_SINGLE_LINE_COMMENT, WHITE_SPACE)) {
            return Indent.getNormalIndent()
        }

        if (elementType == ObjJ_DOT) {
            return Indent.getContinuationIndent()
        }

        if (elementType == ObjJ_ACCESSOR) {
            return Indent.getNormalIndent()
        }

        if (elementType == ObjJ_AT_ACCESSORS) {
            return Indent.getNormalIndent()
        }

        if (parentType == ObjJ_ACCESSOR) {
            return Indent.getNormalIndent()
        }

        if (superParentType == ObjJ_STATEMENT_OR_BLOCK) {
            if (elementType != ObjJ_BLOCK_ELEMENT) {
                return Indent.getNormalIndent()
            }
            return Indent.getNoneIndent()
        }

        if (parent == ObjJ_STATEMENT_OR_BLOCK || superParent == ObjJ_STATEMENT_OR_BLOCK) {
            return if (elementType == ObjJ_OPEN_BRACE || elementType == ObjJ_CLOSE_BRACE || elementType in ObjJTokenSets.INDENT_CHILDREN)
                Indent.getNoneIndent()
            else
                Indent.getNormalIndent()
        }

        if (parentType == ObjJ_COMPARISON_EXPR_PRIME) {
            return Indent.getContinuationIndent()
        }

        if (elementType == ObjJ_RIGHT_EXPR) {
            return if (node.psi.hasParentOfType(ObjJRightExpr::class.java)) {
                Indent.getNoneIndent()
            } else
                Indent.getNormalIndent()
        }

        if (parentType == ObjJ_RIGHT_EXPR) {
            return Indent.getNoneIndent()
        }

        if (elementType == ObjJ_ACCESSOR_PROPERTY) {
            return Indent.getContinuationIndent()
        }

        if (elementType == ObjJ_ACCESSOR_PROPERTY_TYPE) {
            return Indent.getContinuationIndent()
        }

        if (elementType == ObjJ_EQUALS) {
            return Indent.getContinuationIndent()
        }

        if (parentType == ObjJ_TERNARY_EXPR_PRIME) {
            if (prevSiblingType == ObjJ_COLON || elementType == ObjJ_COLON) {
                return Indent.getContinuationIndent()
            }
            return Indent.getNormalIndent()
        }


        if (FormatterUtil.isPrecededBy(node, ObjJ_ASSIGNMENT_OPERATOR, WHITE_SPACE)) {
            return Indent.getContinuationIndent()
        }
        if (FormatterUtil.isPrecededBy(node, ObjJ_MATH_OP, WHITE_SPACE)) {
            return Indent.getNormalIndent()
        }

        if (FormatterUtil.isPrecededBy(node, ObjJ_COMMA, WHITE_SPACE) && elementType != ObjJ_VARIABLE_DECLARATION) {
            return Indent.getContinuationIndent()
        }

        if (FormatterUtil.isPrecededBy(node, ObjJ_OPEN_BRACE, WHITE_SPACE)) {
            return Indent.getNormalIndent()
        }

        if (FormatterUtil.isPrecededBy(node, ObjJ_OPEN_BRACKET, WHITE_SPACE)) {
            return Indent.getNormalIndent()
        }

        if (FormatterUtil.isPrecededBy(node, ObjJ_OPEN_PAREN, WHITE_SPACE)) {
            return Indent.getNormalIndent()
        }

        //Logger.getLogger("#ObjJIndentProcessor").info("Indent-> type: $elementType(${node.text}); parent: $parentType; prev: $prevSiblingType(${prevSibling?.text}); next: $nextSiblingType(${nextSibling?.text}); ")
        return if (elementType == ObjJ_OPEN_BRACE || elementType == ObjJ_CLOSE_BRACE) {
            Indent.getNoneIndent()
        } else Indent.getNoneIndent()

    }

    companion object {
        val EXPRESSIONS = TokenSet.create(ObjJ_EXPR)
        val LOGGER:Logger by lazy {
            Logger.getLogger("#"+ObjJIndentProcessor::class.java.canonicalName)
        }
    }
}